File: packages/ui-flutter/lib/src/color_picker.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import 'theme.dart';

// -- color math (kept identical, formula-for-formula, across all four platforms) --

List<double> kinetixHexToRgba(String hex) {
  final h = hex.replaceFirst('#', '');
  if (h.length == 8) {
    final n = int.parse(h, radix: 16);
    return [
      ((n >> 24) & 0xFF).toDouble(),
      ((n >> 16) & 0xFF).toDouble(),
      ((n >> 8) & 0xFF).toDouble(),
      (n & 0xFF) / 255,
    ];
  }
  final full = h.length == 3 ? h.split('').map((c) => '$c$c').join() : h;
  final n = int.parse(full, radix: 16);
  return [((n >> 16) & 0xFF).toDouble(), ((n >> 8) & 0xFF).toDouble(), (n & 0xFF).toDouble(), 1.0];
}

String kinetixRgbaToHex(double r, double g, double b, double a, {required bool includeAlpha}) {
  String ch(double v) => v.round().clamp(0, 255).toRadixString(16).padLeft(2, '0');
  final base = '#${ch(r)}${ch(g)}${ch(b)}';
  return includeAlpha ? '$base${ch(a * 255)}' : base;
}

List<double> kinetixHsvToRgb(double h, double s, double v) {
  final sN = s / 100;
  final vN = v / 100;
  final c = vN * sN;
  final hh = h / 60;
  final x = c * (1 - (hh % 2 - 1).abs());
  double r = 0, g = 0, b = 0;
  if (hh < 1) {
    r = c;
    g = x;
  } else if (hh < 2) {
    r = x;
    g = c;
  } else if (hh < 3) {
    g = c;
    b = x;
  } else if (hh < 4) {
    g = x;
    b = c;
  } else if (hh < 5) {
    r = x;
    b = c;
  } else {
    r = c;
    b = x;
  }
  final m = vN - c;
  return [(r + m) * 255, (g + m) * 255, (b + m) * 255];
}

List<double> kinetixRgbToHsv(double r, double g, double b) {
  final rN = r / 255, gN = g / 255, bN = b / 255;
  final maxV = math.max(rN, math.max(gN, bN));
  final minV = math.min(rN, math.min(gN, bN));
  final d = maxV - minV;
  double h = 0;
  if (d != 0) {
    if (maxV == rN) {
      h = (gN - bN) / d;
    } else if (maxV == gN) {
      h = (bN - rN) / d + 2;
    } else {
      h = (rN - gN) / d + 4;
    }
    h *= 60;
    if (h < 0) h += 360;
  }
  final s = maxV == 0 ? 0.0 : d / maxV;
  return [h, s * 100, maxV * 100];
}

/// Mirrors `packages/ui/src/components/color-picker.tsx`: a saturation/value
/// square, hue and (optional) alpha rails, a hex field, and swatches.
/// Internal HSV state (not derived from `value` on every build) for the same
/// reason as the web version: saturation 0 or value 0 erase hue information,
/// and re-deriving it every time would make the hue thumb jump to red
/// whenever a drag crosses either edge. `value` only resyncs the internal
/// state when it changes from something other than this widget's own last
/// emission.
///
/// The square and rails are hand-rolled with `GestureDetector` rather than
/// reused from a `Slider` — a 2D gesture has no single-axis primitive to
/// build on, and the rails need per-instance gradient track painting a
/// stock `Slider` doesn't expose. No eyedropper: Flutter has no built-in
/// "sample a pixel anywhere on screen" API the way the web's `EyeDropper`
/// does — a web-only enhancement, same as `KinetixTour`.
class KinetixColorPicker extends StatefulWidget {
  const KinetixColorPicker({
    super.key,
    required this.value,
    required this.onChanged,
    this.alpha = false,
    this.swatches = const [],
  });

  final String value;
  final ValueChanged<String> onChanged;
  final bool alpha;
  final List<String> swatches;

  @override
  State<KinetixColorPicker> createState() => _KinetixColorPickerState();
}

class _KinetixColorPickerState extends State<KinetixColorPicker> {
  late double _h, _s, _v, _a;
  late String _lastEmitted;
  late final TextEditingController _hexController;

  @override
  void initState() {
    super.initState();
    final rgba = kinetixHexToRgba(widget.value);
    final hsv = kinetixRgbToHsv(rgba[0], rgba[1], rgba[2]);
    _h = hsv[0];
    _s = hsv[1];
    _v = hsv[2];
    _a = rgba[3];
    _lastEmitted = widget.value;
    _hexController = TextEditingController(text: widget.value.replaceFirst('#', '').toUpperCase());
  }

  @override
  void didUpdateWidget(covariant KinetixColorPicker oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.value != oldWidget.value && widget.value != _lastEmitted) {
      final rgba = kinetixHexToRgba(widget.value);
      final hsv = kinetixRgbToHsv(rgba[0], rgba[1], rgba[2]);
      setState(() {
        _h = hsv[0];
        _s = hsv[1];
        _v = hsv[2];
        _a = rgba[3];
        _lastEmitted = widget.value;
      });
      _hexController.text = widget.value.replaceFirst('#', '').toUpperCase();
    }
  }

  @override
  void dispose() {
    _hexController.dispose();
    super.dispose();
  }

  void _commit(double h, double s, double v, double a) {
    final rgb = kinetixHsvToRgb(h, s, v);
    final hex = kinetixRgbaToHex(rgb[0], rgb[1], rgb[2], a, includeAlpha: widget.alpha);
    setState(() {
      _h = h;
      _s = s;
      _v = v;
      _a = a;
      _lastEmitted = hex;
    });
    _hexController.text = hex.replaceFirst('#', '').toUpperCase();
    widget.onChanged(hex);
  }

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    final currentColor = HSVColor.fromAHSV(1, _h, (_s / 100).clamp(0, 1).toDouble(), (_v / 100).clamp(0, 1).toDouble()).toColor();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      mainAxisSize: MainAxisSize.min,
      children: [
        LayoutBuilder(
          builder: (context, constraints) {
            void update(Offset local) {
              final ns = (local.dx / constraints.maxWidth * 100).clamp(0, 100).toDouble();
              final nv = (100 - local.dy / constraints.maxHeight * 100).clamp(0, 100).toDouble();
              _commit(_h, ns, nv, _a);
            }

            return GestureDetector(
              onPanDown: (d) => update(d.localPosition),
              onPanUpdate: (d) => update(d.localPosition),
              child: ClipRRect(
                borderRadius: BorderRadius.circular(8),
                child: SizedBox(
                  height: 160,
                  width: double.infinity,
                  child: Stack(
                    children: [
                      ColoredBox(color: HSVColor.fromAHSV(1, _h, 1, 1).toColor()),
                      const DecoratedBox(
                        decoration: BoxDecoration(
                          gradient: LinearGradient(colors: [Colors.white, Colors.transparent]),
                        ),
                      ),
                      const DecoratedBox(
                        decoration: BoxDecoration(
                          gradient: LinearGradient(
                            begin: Alignment.topCenter,
                            end: Alignment.bottomCenter,
                            colors: [Colors.transparent, Colors.black],
                          ),
                        ),
                      ),
                      Positioned(
                        left: constraints.maxWidth * _s / 100 - 7,
                        top: constraints.maxHeight * (1 - _v / 100) - 7,
                        child: Container(
                          width: 14,
                          height: 14,
                          decoration: BoxDecoration(
                            shape: BoxShape.circle,
                            color: currentColor,
                            border: Border.all(color: Colors.white, width: 2),
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            );
          },
        ),
        const SizedBox(height: 12),
        _GradientRail(
          fraction: _h / 360,
          gradient: const LinearGradient(
            colors: [
              Color(0xFFFF0000),
              Color(0xFFFFFF00),
              Color(0xFF00FF00),
              Color(0xFF00FFFF),
              Color(0xFF0000FF),
              Color(0xFFFF00FF),
              Color(0xFFFF0000),
            ],
          ),
          onChanged: (f) => _commit(f * 360, _s, _v, _a),
        ),
        if (widget.alpha) ...[
          const SizedBox(height: 12),
          _GradientRail(
            fraction: _a,
            backgroundColor: const Color(0xFFE5E5E5),
            gradient: LinearGradient(colors: [Colors.transparent, currentColor]),
            onChanged: (f) => _commit(_h, _s, _v, f),
          ),
        ],
        const SizedBox(height: 12),
        Row(
          children: [
            Container(
              width: 32,
              height: 32,
              decoration: BoxDecoration(
                color: currentColor.withValues(alpha: _a),
                borderRadius: BorderRadius.circular(6),
                border: Border.all(color: c.border),
              ),
            ),
            const SizedBox(width: 8),
            Text('#', style: TextStyle(color: c.mutedForeground)),
            Expanded(
              child: TextField(
                controller: _hexController,
                style: const TextStyle(fontFamily: 'monospace', fontSize: 12),
                decoration: const InputDecoration(border: InputBorder.none, isDense: true),
                onSubmitted: (draft) {
                  final pattern = widget.alpha ? RegExp(r'^[0-9a-fA-F]{6}([0-9a-fA-F]{2})?$') : RegExp(r'^[0-9a-fA-F]{6}$');
                  if (pattern.hasMatch(draft)) {
                    widget.onChanged('#${draft.toLowerCase()}');
                  } else {
                    _hexController.text = widget.value.replaceFirst('#', '').toUpperCase();
                  }
                },
              ),
            ),
          ],
        ),
        if (widget.swatches.isNotEmpty) ...[
          const SizedBox(height: 12),
          Wrap(
            spacing: 6,
            runSpacing: 6,
            children: [
              for (final sw in widget.swatches)
                GestureDetector(
                  onTap: () {
                    final rgba = kinetixHexToRgba(sw);
                    final hsv = kinetixRgbToHsv(rgba[0], rgba[1], rgba[2]);
                    _commit(hsv[0], hsv[1], hsv[2], _a);
                  },
                  child: Container(
                    width: 24,
                    height: 24,
                    decoration: BoxDecoration(
                      color: Color(int.parse('FF${sw.replaceFirst('#', '')}', radix: 16)),
                      borderRadius: BorderRadius.circular(4),
                    ),
                  ),
                ),
            ],
          ),
        ],
      ],
    );
  }
}

class _GradientRail extends StatelessWidget {
  const _GradientRail({
    required this.fraction,
    required this.gradient,
    required this.onChanged,
    this.backgroundColor,
  });

  final double fraction;
  final Gradient gradient;
  final Color? backgroundColor;
  final ValueChanged<double> onChanged;

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        void update(Offset local) {
          onChanged((local.dx / constraints.maxWidth).clamp(0, 1).toDouble());
        }

        return GestureDetector(
          onPanDown: (d) => update(d.localPosition),
          onPanUpdate: (d) => update(d.localPosition),
          child: SizedBox(
            width: constraints.maxWidth,
            height: 20,
            child: Stack(
              clipBehavior: Clip.none,
              children: [
                Positioned(
                  left: 0,
                  right: 0,
                  top: 4,
                  height: 12,
                  child: DecoratedBox(
                    decoration: BoxDecoration(
                      color: backgroundColor,
                      gradient: gradient,
                      borderRadius: BorderRadius.circular(6),
                    ),
                  ),
                ),
                Positioned(
                  left: constraints.maxWidth * fraction - 7,
                  top: 3,
                  child: Container(
                    width: 14,
                    height: 14,
                    decoration: const BoxDecoration(
                      shape: BoxShape.circle,
                      color: Colors.white,
                      boxShadow: [BoxShadow(color: Colors.black26, blurRadius: 2)],
                    ),
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/ColorPicker.swift
//
// ColorPicker.swift — KinetixColorPicker.
//
// Mirrors packages/ui/src/components/color-picker.tsx: a saturation/value
// square, hue and (optional) alpha rails, a hex field, and swatches.
// Internal HSV state (not derived from `value` on every render) for the
// same reason as the web version: saturation 0 or value 0 erase hue
// information, and re-deriving it every time would make the hue thumb
// jump to red whenever a drag crosses either edge. `value` only resyncs
// the internal state when it changes from something other than this
// view's own last emission.
//
// The square and rails are hand-rolled with `DragGesture` rather than
// reused from `KinetixSlider` — a 2D gesture has no single-axis primitive
// to build on, and the rails need per-instance gradient track painting
// `KinetixSlider`'s fixed styling doesn't expose. No eyedropper: iOS has
// no public API for sampling a pixel from anywhere on screen the way the
// web's `EyeDropper` does — a web-only enhancement, same as `KinetixTour`.
//

import SwiftUI

// -- color math (kept identical, formula-for-formula, across all four platforms) --

func kinetixHexToRgba(_ hex: String) -> (Double, Double, Double, Double) {
    var h = hex
    if h.hasPrefix("#") { h.removeFirst() }
    if h.count == 8 {
        let n = UInt64(h, radix: 16) ?? 0
        return (
            Double((n >> 24) & 0xFF),
            Double((n >> 16) & 0xFF),
            Double((n >> 8) & 0xFF),
            Double(n & 0xFF) / 255
        )
    }
    let full = h.count == 3 ? h.map { "\($0)\($0)" }.joined() : h
    let n = UInt64(full, radix: 16) ?? 0
    return (Double((n >> 16) & 0xFF), Double((n >> 8) & 0xFF), Double(n & 0xFF), 1)
}

func kinetixRgbaToHex(_ r: Double, _ g: Double, _ b: Double, _ a: Double, includeAlpha: Bool) -> String {
    func ch(_ v: Double) -> String {
        let clamped = max(0, min(255, Int(v)))
        return String(format: "%02x", clamped)
    }
    let base = "#\(ch(r))\(ch(g))\(ch(b))"
    return includeAlpha ? "\(base)\(ch(a * 255))" : base
}

func kinetixHsvToRgb(_ h: Double, _ s: Double, _ v: Double) -> (Double, Double, Double) {
    let sN = s / 100
    let vN = v / 100
    let c = vN * sN
    let hh = h / 60
    let x = c * (1 - abs(hh.truncatingRemainder(dividingBy: 2) - 1))
    var r = 0.0, g = 0.0, b = 0.0
    switch hh {
    case ..<1: (r, g, b) = (c, x, 0)
    case ..<2: (r, g, b) = (x, c, 0)
    case ..<3: (r, g, b) = (0, c, x)
    case ..<4: (r, g, b) = (0, x, c)
    case ..<5: (r, g, b) = (x, 0, c)
    default: (r, g, b) = (c, 0, x)
    }
    let m = vN - c
    return ((r + m) * 255, (g + m) * 255, (b + m) * 255)
}

func kinetixRgbToHsv(_ r: Double, _ g: Double, _ b: Double) -> (Double, Double, Double) {
    let rN = r / 255, gN = g / 255, bN = b / 255
    let maxV = max(rN, gN, bN)
    let minV = min(rN, gN, bN)
    let d = maxV - minV
    var h = 0.0
    if d != 0 {
        if maxV == rN {
            h = (gN - bN) / d
        } else if maxV == gN {
            h = (bN - rN) / d + 2
        } else {
            h = (rN - gN) / d + 4
        }
        h *= 60
        if h < 0 { h += 360 }
    }
    let s = maxV == 0 ? 0 : d / maxV
    return (h, s * 100, maxV * 100)
}

public struct KinetixColorPicker: View {
    @Environment(\.kinetixColors) private var colors

    private let value: String
    private let onChange: (String) -> Void
    private let alpha: Bool
    private let swatches: [String]

    @State private var h: Double
    @State private var s: Double
    @State private var v: Double
    @State private var a: Double
    @State private var lastEmitted: String
    @State private var hexDraft: String

    public init(value: String, onChange: @escaping (String) -> Void, alpha: Bool = false, swatches: [String] = []) {
        self.value = value
        self.onChange = onChange
        self.alpha = alpha
        self.swatches = swatches
        let rgba = kinetixHexToRgba(value)
        let hsv = kinetixRgbToHsv(rgba.0, rgba.1, rgba.2)
        _h = State(initialValue: hsv.0)
        _s = State(initialValue: hsv.1)
        _v = State(initialValue: hsv.2)
        _a = State(initialValue: rgba.3)
        _lastEmitted = State(initialValue: value)
        _hexDraft = State(initialValue: value.replacingOccurrences(of: "#", with: "").uppercased())
    }

    private func commit(_ nh: Double, _ ns: Double, _ nv: Double, _ na: Double) {
        h = nh
        s = ns
        v = nv
        a = na
        let rgb = kinetixHsvToRgb(nh, ns, nv)
        let hex = kinetixRgbaToHex(rgb.0, rgb.1, rgb.2, na, includeAlpha: alpha)
        lastEmitted = hex
        hexDraft = hex.replacingOccurrences(of: "#", with: "").uppercased()
        onChange(hex)
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            GeometryReader { geo in
                ZStack(alignment: .topLeading) {
                    Color(hue: h / 360, saturation: 1, brightness: 1)
                    LinearGradient(colors: [.white, .clear], startPoint: .leading, endPoint: .trailing)
                    LinearGradient(colors: [.clear, .black], startPoint: .top, endPoint: .bottom)
                    Circle()
                        .fill(Color(hue: h / 360, saturation: s / 100, brightness: v / 100))
                        .frame(width: 14, height: 14)
                        .overlay(Circle().stroke(Color.white, lineWidth: 2))
                        .position(x: geo.size.width * s / 100, y: geo.size.height * (1 - v / 100))
                }
                .clipShape(RoundedRectangle(cornerRadius: 8))
                .gesture(
                    DragGesture(minimumDistance: 0).onChanged { drag in
                        let ns = min(100, max(0, drag.location.x / geo.size.width * 100))
                        let nv = min(100, max(0, 100 - drag.location.y / geo.size.height * 100))
                        commit(h, ns, nv, a)
                    }
                )
            }
            .frame(height: 160)

            gradientRail(
                fraction: h / 360,
                track: AnyShapeStyle(AngularGradient(
                    gradient: Gradient(colors: (0...6).map { Color(hue: Double($0) / 6, saturation: 1, brightness: 1) }),
                    center: .center,
                    startAngle: .degrees(0),
                    endAngle: .degrees(360)
                )),
                onChange: { f in commit(f * 360, s, v, a) }
            )

            if alpha {
                gradientRail(
                    fraction: a,
                    track: AnyShapeStyle(Color(white: 0.9)),
                    overlay: LinearGradient(
                        colors: [.clear, Color(hue: h / 360, saturation: s / 100, brightness: v / 100)],
                        startPoint: .leading,
                        endPoint: .trailing
                    ),
                    onChange: { f in commit(h, s, v, f) }
                )
            }

            HStack(spacing: 8) {
                RoundedRectangle(cornerRadius: 6)
                    .fill(Color(hue: h / 360, saturation: s / 100, brightness: v / 100).opacity(a))
                    .frame(width: 32, height: 32)
                    .overlay(RoundedRectangle(cornerRadius: 6).stroke(colors.border, lineWidth: 1))

                HStack(spacing: 2) {
                    Text("#").foregroundStyle(colors.mutedForeground)
                    TextField("", text: $hexDraft)
                        .font(.system(size: 12, design: .monospaced))
                        .textFieldStyle(.plain)
                        .onSubmit {
                            let pattern = alpha ? "^[0-9a-fA-F]{6}([0-9a-fA-F]{2})?$" : "^[0-9a-fA-F]{6}$"
                            if hexDraft.range(of: pattern, options: .regularExpression) != nil {
                                onChange("#\(hexDraft.lowercased())")
                            } else {
                                hexDraft = value.replacingOccurrences(of: "#", with: "").uppercased()
                            }
                        }
                }
            }

            if !swatches.isEmpty {
                HStack(spacing: 6) {
                    ForEach(swatches, id: \.self) { sw in
                        let rgba = kinetixHexToRgba(sw)
                        RoundedRectangle(cornerRadius: 4)
                            .fill(Color(red: rgba.0 / 255, green: rgba.1 / 255, blue: rgba.2 / 255))
                            .frame(width: 24, height: 24)
                            .onTapGesture {
                                let hsv = kinetixRgbToHsv(rgba.0, rgba.1, rgba.2)
                                commit(hsv.0, hsv.1, hsv.2, a)
                            }
                    }
                }
            }
        }
        .onChange(of: value) { newValue in
            guard newValue != lastEmitted else { return }
            let rgba = kinetixHexToRgba(newValue)
            let hsv = kinetixRgbToHsv(rgba.0, rgba.1, rgba.2)
            h = hsv.0
            s = hsv.1
            v = hsv.2
            a = rgba.3
            lastEmitted = newValue
            hexDraft = newValue.replacingOccurrences(of: "#", with: "").uppercased()
        }
    }

    @ViewBuilder
    private func gradientRail(
        fraction: Double,
        track: AnyShapeStyle,
        overlay: LinearGradient? = nil,
        onChange: @escaping (Double) -> Void
    ) -> some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(track)
                if let overlay {
                    Capsule().fill(overlay)
                }
                Circle()
                    .fill(Color.white)
                    .frame(width: 14, height: 14)
                    .overlay(Circle().stroke(Color.white, lineWidth: 2))
                    .shadow(radius: 1)
                    .offset(x: geo.size.width * fraction - 7)
            }
            .frame(height: 12)
            .contentShape(Rectangle())
            .gesture(
                DragGesture(minimumDistance: 0).onChanged { drag in
                    onChange(min(1, max(0, drag.location.x / geo.size.width)))
                }
            )
        }
        .frame(height: 20)
    }
}
